elide-async/src/main/java/com/yahoo/elide/async/service/AsyncQueryThread.java

```diff
@@ -72,35 +81,164 @@ public AsyncQueryThread(AsyncQuery queryObj, User user, Elide elide, QueryRunner
     protected AsyncQueryResult processQuery() throws URISyntaxException, NoHttpResponseException {
 
         ElideResponse response = null;
+        Integer recCount = null;
+
+        Date startTime = new Date();
+
         log.debug("AsyncQuery Object from request: {}", queryObj);
         if (queryObj.getQueryType().equals(QueryType.JSONAPI_V1_0)) {
             MultivaluedMap<String, String> queryParams = getQueryParams(queryObj.getQuery());
             log.debug("Extracted QueryParams from AsyncQuery Object: {}", queryParams);
             response = elide.get(getPath(queryObj.getQuery()), queryParams, user, apiVersion);
             log.debug("JSONAPI_V1_0 getResponseCode: {}, JSONAPI_V1_0 getBody: {}",
                     response.getResponseCode(), response.getBody());
+
+            if (response.getResponseCode() == 200) {
+                recCount = calculateRecordsJSON(response.getBody());
+            }
         }
         else if (queryObj.getQueryType().equals(QueryType.GRAPHQL_V1_0)) {
             response = runner.run(queryObj.getQuery(), user);
             log.debug("GRAPHQL_V1_0 getResponseCode: {}, GRAPHQL_V1_0 getBody: {}",
                     response.getResponseCode(), response.getBody());
+
+            if (response.getResponseCode() == 200) {
+                String tableName = getTableNameFromQuery(queryObj.getQuery());
+                recCount = calculateRecordsGRAPHQL(response.getBody(), tableName);
+            }
         }
         if (response == null) {
             throw new NoHttpResponseException("Response for request returned as null");
         }
 
+        long timeElapsed = ((new Date()).getTime() - startTime.getTime());
+        long asyncAfterMilliSeconds = queryObj.getAsyncAfterSeconds() * 1000;
+
         // Create AsyncQueryResult entry for AsyncQuery
 
         queryResultObj = new AsyncQueryResult();
         queryResultObj.setHttpStatus(response.getResponseCode());
-        queryResultObj.setResponseBody(response.getBody());
         queryResultObj.setContentLength(response.getBody().length());
-        queryResultObj.setResultType(ResultType.EMBEDDED);
+        queryResultObj.setRecordCount(recCount);
         queryResultObj.setCompletedOn(new Date());
 
+        if (timeElapsed - asyncAfterMilliSeconds > 0) {
+            String jsonToCSV = convertJsonToCSV(response.getBody());
+            URL url = null;
+            if (queryObj.getResultFormatType() == ResultFormatType.CSV) {
+                url = resultStorageEngine.storeResults(queryObj.getId(), jsonToCSV);
```

Comment:
we can create an AsyncResultStorage Object with Id and Blob. and pass that as argument to storeResults method.